In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_parquet('green_tripdata_2022-01.parquet')

In [6]:
df.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         2  2022-01-01 00:14:21   2022-01-01 00:15:33                  N   
1         1  2022-01-01 00:20:55   2022-01-01 00:29:38                  N   
2         1  2022-01-01 00:57:02   2022-01-01 01:13:14                  N   
3         2  2022-01-01 00:07:42   2022-01-01 00:15:57                  N   
4         2  2022-01-01 00:07:50   2022-01-01 00:28:52                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            42            42              1.0           0.44   
1         1.0           116            41              1.0           2.10   
2         1.0            41           140              1.0           3.70   
3         1.0           181           181              1.0           1.69   
4         1.0            33           170              1.0           6.26   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount ehail_fee  \
0          3.5   0.50      0.5        0.00           0.0      None   
1          9.5   0.50      0.5        0.00           0.0      None   
2         14.5   3.25      0.5        4.60           0.0      None   
3          8.0   0.50      0.5        0.00           0.0      None   
4         22.0   0.50      0.5        5.21           0.0      None   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          4.80           2.0        1.0   
1                    0.3         10.80           2.0        1.0   
2                    0.3         23.15           1.0        1.0   
3                    0.3          9.30           2.0        1.0   
4                    0.3         31.26           1.0        1.0   

   congestion_surcharge  
0                  0.00  
1                  0.00  
2                  2.75  
3                  0.00  
4                  2.75

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62495 entries, 0 to 62494
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               62495 non-null  int64         
 1   lpep_pickup_datetime   62495 non-null  datetime64[ns]
 2   lpep_dropoff_datetime  62495 non-null  datetime64[ns]
 3   store_and_fwd_flag     56200 non-null  object        
 4   RatecodeID             56200 non-null  float64       
 5   PULocationID           62495 non-null  int64         
 6   DOLocationID           62495 non-null  int64         
 7   passenger_count        56200 non-null  float64       
 8   trip_distance          62495 non-null  float64       
 9   fare_amount            62495 non-null  float64       
 10  extra                  62495 non-null  float64       
 11  mta_tax                62495 non-null  float64       
 12  tip_amount             62495 non-null  float64       
 13  t

In [8]:
urls = ['https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-01.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-02.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-03.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-04.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-05.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-06.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-07.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-08.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-09.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-10.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-11.parquet',
        'https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-12.parquet',
        ]

In [14]:
full_df = pd.DataFrame()
for parquet_url in urls:
    df = pd.read_parquet(parquet_url)
    #print(len(df))
    full_df = pd.concat([full_df, df])
    #print(len(full_df))

62495
62495
69399
131894
78537
210431
76136
286567
76891
363458
73718
437176
64192
501368
65929
567297
69031
636328
69322
705650
62313
767963
72439
840402


In [15]:
full_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 840402 entries, 0 to 72438
Data columns (total 20 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               840402 non-null  int64         
 1   lpep_pickup_datetime   840402 non-null  datetime64[ns]
 2   lpep_dropoff_datetime  840402 non-null  datetime64[ns]
 3   store_and_fwd_flag     750062 non-null  object        
 4   RatecodeID             750062 non-null  float64       
 5   PULocationID           840402 non-null  int64         
 6   DOLocationID           840402 non-null  int64         
 7   passenger_count        750062 non-null  float64       
 8   trip_distance          840402 non-null  float64       
 9   fare_amount            840402 non-null  float64       
 10  extra                  840402 non-null  float64       
 11  mta_tax                840402 non-null  float64       
 12  tip_amount             840402 non-null  float64   

In [16]:
full_df['lpep_pickup_datetime'] = pd.to_datetime(full_df['lpep_pickup_datetime'])
full_df['lpep_dropoff_datetime'] = pd.to_datetime(full_df['lpep_dropoff_datetime'])
full_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 840402 entries, 0 to 72438
Data columns (total 20 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               840402 non-null  int64         
 1   lpep_pickup_datetime   840402 non-null  datetime64[ns]
 2   lpep_dropoff_datetime  840402 non-null  datetime64[ns]
 3   store_and_fwd_flag     750062 non-null  object        
 4   RatecodeID             750062 non-null  float64       
 5   PULocationID           840402 non-null  int64         
 6   DOLocationID           840402 non-null  int64         
 7   passenger_count        750062 non-null  float64       
 8   trip_distance          840402 non-null  float64       
 9   fare_amount            840402 non-null  float64       
 10  extra                  840402 non-null  float64       
 11  mta_tax                840402 non-null  float64       
 12  tip_amount             840402 non-null  float64   

## Question 1:
Question 1: What is count of records for the 2022 Green Taxi Data??

> SELECT  COUNT(*)

> FROM `de-zoomcamp-411910.green_taxi_2022.gree_taxi_2022_table` 

* 65,623,481
* **840,402**
* 1,936,423
* 253,647

## Question 2:

Write a query to count the distinct number of PULocationIDs for the entire dataset on both the tables.
What is the estimated amount of data that will be read when this query is executed on the External Table and the Table?

> SELECT  COUNT(*)

> FROM `de-zoomcamp-411910.green_taxi_2022.gree_taxi_2022_table` 

> GROUP BY PULocationID

* **0 MB for the External Table and 6.41MB for the Materialized Table**
* 18.82 MB for the External Table and 47.60 MB for the Materialized Table
* 0 MB for the External Table and 0MB for the Materialized Table
* 2.14 MB for the External Table and 0MB for the Materialized Table

## Question 3:
How many records have a fare_amount of 0?

> SELECT  COUNT(*)

> FROM `de-zoomcamp-411910.green_taxi_2022.gree_taxi_2022_table` 

> WHERE fare_amount = 0


* 12,488
* 128,219
* 112
* **1,622**

## Question 4:
What is the best strategy to make an optimized table in Big Query if your query will always order the results by PUlocationID and filter based on lpep_pickup_datetime? (Create a new table with this strategy)

> CREATE TABLE green_taxi_2022.clustered_table

> PARTITION BY lpep_pickup_datetime

> CLUSTER BY PUlocationID
  
> AS (SELECT * FROM de-zoomcamp-411910.green_taxi_2022.green_taxi_2022_table);
  

* **Cluster on lpep_pickup_datetime Partition by PUlocationID**
* Partition by lpep_pickup_datetime Cluster on PUlocationID
* Partition by lpep_pickup_datetime and Partition by PUlocationID
* Cluster on by lpep_pickup_datetime and Cluster on PUlocationID

## Question 5:

Write a query to retrieve the distinct PULocationID between lpep_pickup_datetime 06/01/2022 and 06/30/2022 (inclusive)
Use the materialized table you created earlier in your from clause and note the estimated bytes. Now change the table in the from clause to the partitioned table you created for question 4 and note the estimated bytes processed. What are these values?
Choose the answer which most closely matches.

* 22.82 MB for non-partitioned table and 647.87 MB for the partitioned table
* **12.82 MB for non-partitioned table and 1.12 MB for the partitioned table**
* 5.63 MB for non-partitioned table and 0 MB for the partitioned table
* 10.31 MB for non-partitioned table and 10.31 MB for the partitioned table

## Question 6:
Where is the data stored in the External Table you created?

* Big Query
* **GCP Bucket**
* Big Table
* Container Registry

## Question 7:
It is best practice in Big Query to always cluster your data:

* True
* **False**